src/esp/assets/ResourceManager.h

```diff
@@ -807,6 +818,54 @@ class ResourceManager {
    */
   void loadMaterials(Importer& importer, LoadedAssetData& loadedAssetData);
 
+  /**
+   * @brief Get the appropriate the @ref
+   * esp::metadata::attributes::ObjectInstanceShaderType to use to render the
+   * passed @p material based on specification in passed @p info or the material
+   * itself.
+   * @param info The asset info describing the asset whose material is being
+   * rendered.
+   * @return the @ref esp::metadata::attributes::ObjectInstanceShaderType to use
+   * to render the material.
+   */
+  ShaderTypeEnum getMaterialShaderType(const AssetInfo& info) const;
+
+  /**
+   * @brief Boolean check if @p typeToCheck aligns with passed types explicitly
+   * specified, or type in material
+   * @param typeToCheck The ShaderTypeEnum value beign queried for.
+   * @param materialData The material whose type we are verifying against
+   * @param verificationType The ShaderTypeEnum we are verifying against
+   * @param mnVerificationType The @ref Mn::Trade::MaterialType bitflag that the
+   * passed material's specified type is being verified against.
+   * @return Whether or not the passed @p typeToCheck matches the passed
+   * criteria.
+   */
+  bool checkForPassedShaderType(
+      const ShaderTypeEnum typeToCheck,
+      const Cr::Containers::Optional<Mn::Trade::MaterialData>& materialData,
```

Comment:
Why pass an argument through an `Optional` if you don't allow / don't expect `NullOpt` values? In that case I'd take just a `const Mn::Trade::MaterialData&`. The caller then passes `*foo` instead of `foo` (and ensures it's not a null optional, of course).

This has an equivalent semantics to a reference vs a pointer -- if the API allows `nullptr` values, a pointer is fine, otherwise it should be a reference, and the responsibility of ensuring non-nullness lies then on the caller.

I see the same being done for a lot more functions (`buildFlatShadedMaterialData()`, ...), since they were taking a const reference before I don't think the Optional has any use there either.

Reply:
Thanks for checking things out @mosra, as always, and for pointing out these inefficiencies.  In general, when I'm refactoring existing code I try to keep my changes as minimal as possible, so I may propagate design choices that were sub-optimal.  I'll address them as I improve this WIP.